Add support for raw-dylib with stdcall, fastcall functions

ricobbe opened the pull request:
Next stage of work for #58713: allow `extern "stdcall"` and `extern "fastcall"` with `#[link(kind = "raw-dylib")]`.

I've deliberately omitted support for vectorcall, as that doesn't currently work, and I wanted to get this out for review.  (I haven't really investigated the vectorcall failure much yet, but at first (very cursory) glance it appears that the problem is elsewhere.)

rust-highfive commented:
r? @matthewjasper

(rust-highfive has picked a reviewer for you, use r? to override)

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-10`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/2854395972) [(plain)](https://github.com/rust-lang/rust/commit/69c093f979dda7c15b0b41afb788fa02c0bcef1e/checks/2854395972/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Suite("src/test/run-make") not skipped for "bootstrap::test::RunMake" -- not in ["src/tools/tidy"]
Check compiletest suite=run-make mode=run-make (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)

running 30 tests
Some tests failed in compiletest suite=run-make mode=run-make host=x86_64-unknown-linux-gnu target=x86_64-unknown-linux-gnu
iiiiiiiiiiiiiii.F.............


---- [run-make] run-make/raw-dylib-stdcall stdout ----
error: make failed
status: exit status: 2
command: "make"
stdout:
stdout:
------------------------------------------
cc -ffunction-sections -fdata-sections -fPIC -m64 -c -o "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make/raw-dylib-stdcall/raw-dylib-stdcall"/extern.obj extern.c
Makefile:8: recipe for target 'all' failed
------------------------------------------
stderr:
------------------------------------------
extern.c:9:1: warning: return type defaults to 'int' [-Wimplicit-int]
extern.c:9:1: warning: return type defaults to 'int' [-Wimplicit-int]
 __declspec(dllexport) void __stdcall stdcall_fn_1(int i) {
extern.c: In function '__declspec':
extern.c: In function '__declspec':
extern.c:9:38: error: expected '=', ',', ';', 'asm' or '__attribute__' before 'stdcall_fn_1'
 __declspec(dllexport) void __stdcall stdcall_fn_1(int i) {
                                      ^~~~~~~~~~~~
extern.c:14:1: error: expected declaration specifiers before '__declspec'
 __declspec(dllexport) void __stdcall stdcall_fn_2(uint8_t i, float f) {
 ^~~~~~~~~~
extern.c:19:1: error: expected declaration specifiers before '__declspec'
 __declspec(dllexport) void __stdcall stdcall_fn_3(double d) {
 ^~~~~~~~~~
extern.c:24:1: error: expected declaration specifiers before '__declspec'
 __declspec(dllexport) void __stdcall stdcall_fn_4(uint8_t i, uint8_t j, float f) {
 ^~~~~~~~~~
extern.c:29:1: error: expected declaration specifiers before '__declspec'
 __declspec(dllexport) void __stdcall stdcall_fn_5(struct S s, int i) {
 ^~~~~~~~~~
extern.c:34:1: error: expected declaration specifiers before '__declspec'
 __declspec(dllexport) void __stdcall stdcall_fn_6(struct S* s) {
 ^~~~~~~~~~
extern.c:9:1: warning: type of 'dllexport' defaults to 'int' [-Wimplicit-int]
 __declspec(dllexport) void __stdcall stdcall_fn_1(int i) {
 ^~~~~~~~~~
extern.c:41:1: error: expected '{' at end of input
 ^
 ^
make: *** [all] Error 1
------------------------------------------




failures:
    [run-make] run-make/raw-dylib-stdcall

test result: FAILED. 14 passed; 1 failed; 15 ignored; 0 measured; 0 filtered out; finished in 2.88s



command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/compiletest" "--compile-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" "--run-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "--rustc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "--rustdoc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustdoc" "--src-base" "/checkout/src/test/run-make" "--build-base" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/run-make" "--stage-id" "stage2-x86_64-unknown-linux-gnu" "--suite" "run-make" "--mode" "run-make" "--target" "x86_64-unknown-linux-gnu" "--host" "x86_64-unknown-linux-gnu" "--llvm-filecheck" "/usr/lib/llvm-10/bin/FileCheck" "--nodejs" "/usr/bin/node" "--host-rustcflags" "-Crpath -O -Cdebuginfo=0  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--target-rustcflags" "-Crpath -O -Cdebuginfo=0  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--docck-python" "/usr/bin/python3" "--lldb-python" "/usr/bin/python3" "--gdb" "/usr/bin/gdb" "--quiet" "--llvm-version" "10.0.0" "--llvm-components" "aarch64 aarch64asmparser aarch64codegen aarch64desc aarch64disassembler aarch64info aarch64utils aggressiveinstcombine all all-targets amdgpu amdgpuasmparser amdgpucodegen amdgpudesc amdgpudisassembler amdgpuinfo amdgpuutils analysis arm armasmparser armcodegen armdesc armdisassembler arminfo armutils asmparser asmprinter avr avrasmparser avrcodegen avrdesc avrdisassembler avrinfo binaryformat bitreader bitstreamreader bitwriter bpf bpfasmparser bpfcodegen bpfdesc bpfdisassembler bpfinfo cfguard codegen core coroutines coverage debuginfocodeview debuginfodwarf debuginfogsym debuginfomsf debuginfopdb demangle dlltooldriver dwarflinker engine executionengine frontendopenmp fuzzmutate globalisel hexagon hexagonasmparser hexagoncodegen hexagondesc hexagondisassembler hexagoninfo instcombine instrumentation interpreter ipo irreader jitlink lanai lanaiasmparser lanaicodegen lanaidesc lanaidisassembler lanaiinfo libdriver lineeditor linker lto mc mca mcdisassembler mcjit mcparser mips mipsasmparser mipscodegen mipsdesc mipsdisassembler mipsinfo mirparser msp430 msp430asmparser msp430codegen msp430desc msp430disassembler msp430info native nativecodegen nvptx nvptxcodegen nvptxdesc nvptxinfo objcarcopts object objectyaml option orcerror orcjit passes perfjitevents powerpc powerpcasmparser powerpccodegen powerpcdesc powerpcdisassembler powerpcinfo profiledata remarks riscv riscvasmparser riscvcodegen riscvdesc riscvdisassembler riscvinfo riscvutils runtimedyld scalaropts selectiondag sparc sparcasmparser sparccodegen sparcdesc sparcdisassembler sparcinfo support symbolize systemz systemzasmparser systemzcodegen systemzdesc systemzdisassembler systemzinfo tablegen target textapi transformutils vectorize webassembly webassemblyasmparser webassemblycodegen webassemblydesc webassemblydisassembler webassemblyinfo windowsmanifest x86 x86asmparser x86codegen x86desc x86disassembler x86info x86utils xcore xcorecodegen xcoredesc xcoredisassembler xcoreinfo xray" "--system-llvm" "--llvm-bin-dir" "/usr/lib/llvm-10/bin" "--cc" "cc" "--cxx" "c++" "--cflags" "-ffunction-sections -fdata-sections -fPIC -m64" "--ar" "ar" "--adb-path" "adb" "--adb-test-dir" "/data/tmp/work" "--android-cross-path" "" "--channel" "nightly" "--color" "always"


failed to run: /checkout/obj/build/bootstrap/debug/bootstrap --stage 2 test --exclude src/tools/tidy
Build completed unsuccessfully in 0:30:33
```

</details>

petrochenkov commented on compiler/rustc_metadata/src/native_libs.rs:
```diff
@@ -396,4 +384,48 @@ impl Collector<'tcx> {
+                }
+            }
+        } else {
+            DllCallingConvention::C
```
This logic is going to normalize arbitrary irrelevant ABIs (e.g. `extern "msp430-interrupt"`) to `extern "C"`.

Also see some discussion on accepting ABIs on various targets in https://github.com/rust-lang/rust/pull/86231.

ricobbe replied:
I read through the comments on the linked PR, and I'm a bit confused about the desired change here.  Should I signal an error on other cpu architectures for anything other than C, cdecl, stdcall, fastcall, and maybe vectorcall?  Or just signal an error for anything except C?  (MSVC ignores the other calling conventions on those platforms, but it's not clear that we should follow suit.)

It's probably worth pointing out that this function should only ever run on windows-msvc targets; I'll add a comment (or, better, an assertion) to that effect.

ricobbe replied:
Also, I notice that rustc supports an i586-pc-windows-msvc target.  Should we support stdcall & fastcall there also?

petrochenkov replied:
>Also, I notice that rustc supports an i586-pc-windows-msvc target. Should we support stdcall & fastcall there also?

It should be automatically supported, `i586` is basically the same target as `i686` just with a different base CPU (controlling the set of available CPU features), see e.g. the target definition in `compiler\rustc_target\src\spec\i586_pc_windows_msvc.rs`.

petrochenkov replied:
>I read through the comments on the linked PR, and I'm a bit confused about the desired change here.

I'm not entirely sure yet.
I suspect that MSVC supported the "legacy" conventions on x64 for compatibility reasons, to ease portability of huge amount of pre-existing code.
We don't have such a problem, so for us it would be better to reject all the "legacy" conventions on x64.

petrochenkov replied:
Let's accept only `Abi::C` and `Abi::Win64` here for a start, and produce the "ABI not supported" error for anything else.

petrochenkov replied:
(Note that vectorcall is a thing on x64, and it's different from the default C/win64, and I have no idea how it's mangled there, but this PR doesn't support vectorcall, so it's not a problem right now.)

ricobbe replied:
Re vectorcall: interesting; I wasn't aware that was a thing on 64-bit.  Good to know!

For context, I decided not to support vectorcall on 32-bit because there appear to be deeper problems with it than just getting the name mangling correct.  Making the obvious modification here to allow vectorcall and running the PR's run-make test (with, effectively, `s/stdcall/vectorcall/g;`) didn't work, but in an interesting way: it all links and runs, but the output indicates that the called function isn't seeing the same parameter values that the calling function provides.  So my hunch is that there are other issues with vectorcall in rustc that are outside the scope of this PR.  My primary motivation with the raw-dylib work is facilitating interop with the Windows API, and effectively all of those functions are stdcall on x86-32, so it just didn't seem worth it to dig into the vectorcall issues further.

ricobbe replied:
Also added `Abi::System` for x86_64, as that appears to be a synonym for `Abi::C` (at least on x86_64) that devs will expect to be able to use.

petrochenkov commented on compiler/rustc_metadata/src/native_libs.rs:
```diff
@@ -396,4 +384,48 @@ impl Collector<'tcx> {
+                .map_bound(|slice| self.tcx.mk_type_list(slice.iter())),
+        );
+
+        // stdcall on i686 never allocates less than 32 bits for an argument.
```
Is this true for fastcall as well?

(Also I suspect that we already have the `i686_arg_list_size` logic somewhere in rustc, but can't check right now.)

ricobbe replied:
It appears to be true: running run-make/raw-dylib-stdcall (replacing `stdcall` with `fastcall` throughout) does work.  As per comment below, I'll add tests for this case, and I'll update the comment.

And I wouldn't be at all surprised to learn that `i686_arg_list_size` is duplicate logic, but after digging around myself and consulting with various people, I was unable to find existing code that does the same thing.  If we can find the existing function, I'll be more than happy to use that instead.

petrochenkov replied:
The closest logic that I could find is `compiler\rustc_target\src\abi\call\x86.rs`, but it looks like the mangling logic doesn't care what is passed in registers or somewhere else, it just sums up the sizes.

ricobbe replied:
You are correct that the mangling logic doesn't differentiate between arguments passed on the stack and arguments passed in registers.  I haven't seen that documented anywhere, but the fact that this PR's run-make test passes provides pretty solid evidence for it.  😃 

I have a vague recollection that the stdcall convention requires the caller to allocate but not initialize stack space for arguments passed in registers, in order to give the called function a place to spill those registers if necessary, but I could be dreaming that up.  Either way, I think we've addressed this point -- if we find a better pre-existing function, then we can start another conversation and I'll address it then.

petrochenkov replied:
Ugh, I looked through some code in LLVM and found at least two instances of this logic:
- `addByteCountSuffix` in `llvm-project\llvm\lib\IR\Mangler.cpp`  - aligns to 32 bit (or pointer size in general) instead of using it as a minimum like this PR, comment (incorrectly?) says that the result is a number of words rather than the number of bytes.
- `MangleContext::mangleName` in `llvm-project\clang\lib\AST\Mangle.cpp` - seems to do the same thing, but with some messy divisions and multiplications to get the number of bytes (again) instead of the number of words.

Looks like the alignment vs minimum mismatch is a thing to fix.

ricobbe replied:
Yeah, I saw the comment in `addByteCountSuffix`, or one just like it.  I can only assume, based on experience, that the use of "words" there is incorrect -- using size in words instead of bytes in the mangled names results in an import library that doesn't link correctly against the DLL produced by MSVC.  The details of stdcall that I've been able to find at docs.microsoft.com are suspiciously silent on this point, so I kinda have to take the empirical results as definitive.  I'd personally prefer to have better clarity in the spec on this point, but 🤷 .

ricobbe replied:
> Looks like the alignment vs minimum mismatch is a thing to fix.

Well, after further experimentation, I agree that `max(4, layout.size.bytes_usize())` is not correct, and I've updated this, although I haven't pushed yet.

Based on what I've observed, it seems that stdcall expects that the space for a given argument is the smallest multiple of 4 bytes into which that argument can fit; this is true even for types like `f64` that might under other circumstances be aligned to the nearest 8-byte boundary.

The alignment information that I get from the `Layout` struct doesn't really help here.  As one example, the `Layout` value for `u8` indicates an alignment of 1 byte (both abi and preferred), which doesn't match stdcall: `fn f(x: u8, y: u8)` has an argument size of 8 bytes, not 2.

Unfortunately, messy code is probably unavoidable -- best I've got so far is `(layout.size.bytes_usize() + 3) & !3`, with a comment explaining what's going on.

Beyond that, is there anything else you want to see here?

petrochenkov replied:
>Beyond that, is there anything else you want to see here?

No, the `min` was the only thing that looked incorrect here.

petrochenkov commented on compiler/rustc_codegen_ssa/src/back/link.rs:
```diff
@@ -550,30 +550,23 @@ fn collate_raw_dylibs(used_libraries: &[NativeLib]) -> Vec<(String, Vec<DllImpor
-    // ordinals.
+    // Rustc already signals an error if we have two imports with the same name but different
+    // calling conventions (or function signatures), so we don't have pay attention to those
+    // when ordering.
```
Is there a test for this case?

ricobbe replied:
If both definitions are in the same scope, it results in E0428 ("the name `mumble` is defined multiple times").  I wasn't able to find a test for the case where the definitions have different calling conventions, but [test/ui/issues/issue-28472.rs](https://github.com/rust-lang/rust/blob/master/src/test/ui/issues/issue-28472.rs) comes close; it's two `extern "C"` functions with the same name and signature.

If the definitions are in different blocks (say, one inside a function and one at module top-level), then rustc just issues a lint warning: "`mumble` redeclared with a different signature;" this is covered by [test/ui/lint/clashing-extern-function.rs](https://github.com/rust-lang/rust/blob/master/src/test/ui/lint/clashing-extern-fn.rs).  Do we need to upgrade this to an error?

petrochenkov replied:
I think we shouldn't touch the lint, but report a new hard error here if mismatching declarations are used with raw-dylib feature specifically.

petrochenkov commented on compiler/rustc_codegen_ssa/src/back/link.rs:
```diff
@@ -550,30 +550,23 @@ fn collate_raw_dylibs(used_libraries: &[NativeLib]) -> Vec<(String, Vec<DllImpor
-                }
+        .map(|(lib_name, import_set)| {
+            let mut imports = Vec::from_iter(import_set.into_iter());
+            imports.sort_unstable_by(|a: &DllImport, b: &DllImport| {
```
```suggestion
            imports.sort_unstable_by_key(|a: &DllImport, b: &DllImport| {
```
is usually more convenient in cases like this.

petrochenkov commented on compiler/rustc_codegen_llvm/src/back/archive.rs:
```diff
@@ -391,6 +391,21 @@ impl<'a> LlvmArchiveBuilder<'a> {
     }
+
+    fn i686_decorated_name(import: &DllImport) -> CString {
+        let name = import.name.to_string();
```
```suggestion
        let name = import.name;
```
Symbols can be `format!`ed without allocating a new string.

petrochenkov commented on compiler/rustc_codegen_llvm/src/back/archive.rs:
```diff
@@ -209,7 +209,7 @@ impl<'a> ArchiveBuilder<'a> for LlvmArchiveBuilder<'a> {
-                |import: &DllImport| CString::new(format!("_{}", import.name.to_string())).unwrap()
+                LlvmArchiveBuilder::i686_decorated_name
             } else {
                 |import: &DllImport| CString::new(import.name.to_string()).unwrap()
```
Nit: it's a bit weird to return two closures from if/else, it's probably better to use one closure and process the `arch == "x86"` condition inside it.

petrochenkov replied:
I meant `let import_name_vector: Vec<CString> = dll_imports.iter().map(new_single_function_for_both_64_bit_and_32_bit targets).collect();`

ricobbe replied:
Changed; I was just lifting the architecture test out of the loop since it's invariant, but I can't imagine that perf matters that much here.

petrochenkov commented on src/test/run-make/raw-dylib-stdcall/lib.rs:
```diff
@@ -0,0 +1,29 @@
+}
+
+#[link(name = "extern", kind = "raw-dylib")]
+extern "stdcall" {
```
Other newly supported ABIs also need test cases.

ricobbe replied:
Added.

One question: I haven't changed the makefile, but there seems to be something wrong with my header command.  Could you have a look and tell me what I'm missing?  When I run the test with
```
x.py test -j 4 --build=i686-pc-windows-msvc src\test\run-make\raw-dylib-alt-calling-convention
```
it ignores the test, and I'm not sure why.  Deleting line 3 of the makefile allows it to run with the same command, so I've clearly gotten something wrong somewhere.

ricobbe replied:
Fixed; apparently it needs to be `# only-i686-pc-windows-msvc` -- the "pc" is crucial, for some reason.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/2898236613) [(plain)](https://github.com/rust-lang/rust/commit/99a0dce4e7c7a1e9b22d19c2c8f33156c879294f/checks/2898236613/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Successfully built ad5ef217f0ff
Successfully tagged rust-ci:latest
Built container sha256:ad5ef217f0fffcd74557f8d52f2833e59702db59656f36546341974c47b16bb7
Uploading finished image to https://ci-caches.rust-lang.org/docker/3ccdc63d9d4b09f940536f268273cee5fff0a2be1dc7e629c4be003d524f779b5660e37b0d2b307b5801865dae2862c25a4c07afffb0a92ab35aed41de335d50
upload failed: - to s3://rust-lang-ci-sccache2/docker/3ccdc63d9d4b09f940536f268273cee5fff0a2be1dc7e629c4be003d524f779b5660e37b0d2b307b5801865dae2862c25a4c07afffb0a92ab35aed41de335d50 Unable to locate credentials
[CI_JOB_NAME=mingw-check]
---
configure: rust.channel         := nightly
configure: rust.debug-assertions := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
---
skip untracked path cpu-usage.csv during rustfmt invocations
skip untracked path src/doc/book/ during rustfmt invocations
skip untracked path src/doc/rust-by-example/ during rustfmt invocations
skip untracked path src/llvm-project/ during rustfmt invocations
Diff in /checkout/compiler/rustc_codegen_llvm/src/back/archive.rs at line 206:
         // All import names are Rust identifiers and therefore cannot contain \0 characters.
         // FIXME: when support for #[link_name] implemented, ensure that import.name values don't
         // have any \0 characters
-        let import_name_vector: Vec<CString> =
-            if self.config.sess.target.arch == "x86" {
-                dll_imports.iter().map(LlvmArchiveBuilder::i686_decorated_name).collect()
-                dll_imports
-                    .iter()
-                    .iter()
-                    .map(|import: &DllImport| CString::new(import.name.to_string()).unwrap())
-                    .collect()
-            };
+        let import_name_vector: Vec<CString> = if self.config.sess.target.arch == "x86" {
+            dll_imports.iter().map(LlvmArchiveBuilder::i686_decorated_name).collect()
+            dll_imports
+                .iter()
+                .iter()
+                .map(|import: &DllImport| CString::new(import.name.to_string()).unwrap())
+                .collect()
 
 
         let output_path_z = rustc_fs_util::path_to_c_string(&output_path);
 
Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/rustfmt" "--config-path" "/checkout" "--edition" "2018" "--unstable-features" "--skip-children" "--check" "/checkout/library/test/src/stats.rs" "/checkout/compiler/rustc_codegen_llvm/src/type_of.rs" "/checkout/library/test/src/types.rs" "/checkout/compiler/rustc_codegen_llvm/src/back/archive.rs" "/checkout/compiler/rustc_codegen_llvm/src/back/lto.rs" "/checkout/compiler/rustc_codegen_llvm/src/back/profiling.rs" "/checkout/compiler/rustc_codegen_llvm/src/back/write.rs" "/checkout/compiler/rustc_codegen_llvm/src/debuginfo/utils.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
Build completed unsuccessfully in 0:00:14
```

</details>

petrochenkov commented on src/test/run-make/raw-dylib-alt-calling-convention/Makefile:
```diff
@@ -0,0 +1,18 @@
+# Test the behavior of #[link(.., kind = "raw-dylib")] with alternative calling conventions.
+
+# only-i686-pc-windows-msvc
```
I suppose this should be testable on all MSVC targets since we are normalizing stdcall/fastcall/etc to C on everything except for `i686`?
(This kind of depends on decisions in https://github.com/rust-lang/rust/pull/86231 again.)

ricobbe replied:
Well, the test passes when run in x86_64-pc-windows-msvc, though I'd like additional clarity on what we want to do with the various calling conventions on non-i686(/i586) platforms before relaxing the `only` header command.

petrochenkov replied:
Ok, let's leave this as is in accordance with https://github.com/rust-lang/rust/pull/86419#discussion_r660780299.

petrochenkov commented on compiler/rustc_codegen_ssa/src/back/link.rs:
```diff
@@ -550,30 +550,21 @@ fn collate_raw_dylibs(used_libraries: &[NativeLib]) -> Vec<(String, Vec<DllImpor
+            (lib_name, imports)
         })
         .collect::<Vec<_>>();
     result.sort_unstable_by(|a: &(String, Vec<DllImport>), b: &(String, Vec<DllImport>)| {
```
```suggestion
    result.sort_unstable_by_key(...
```

ricobbe replied:
Yeah, I've tried to get that to work, but I can't make the lifetimes work (unless I copy the strings, but that seems wasteful).

```
573 |     result.sort_unstable_by_key(|x: &(String, Vec<DllImport>)| &x.0);
    |                                     -                        - ^^^^ returning this value requires that `'1` must outlive `'2`
    |                                     |                        |
    |                                     |                        return type of closure is &'2 std::string::String
    |                                     let's call the lifetime of this reference `'1`
```

Lifting the closure to a named function with lifetime parameters doesn't work either:
```
    fn extract_key<'a, 'b: 'a>(x: &'b (String, Vec<DllImport>)) -> &'a str { &x.0 }

    result.sort_unstable_by_key(extract_key);
```
results in an error saying only "higher-ranked subtype error" at the call site.

petrochenkov replied:
Ok.

petrochenkov commented:
r? @petrochenkov

ricobbe commented:
@rustbot label -S-waiting-on-author +S-waiting-on-review

petrochenkov commented on compiler/rustc_metadata/src/native_libs.rs:
```diff
@@ -396,4 +384,55 @@ impl Collector<'tcx> {
+                .map_bound(|slice| self.tcx.mk_type_list(slice.iter())),
+        );
+
+        argument_types.iter().fold(0, |accum: usize, ty: Ty<'_>| {
```
It would be simpler to use `Iterator::sum` https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.sum instead of `fold` here.

petrochenkov commented on compiler/rustc_codegen_ssa/src/back/link.rs:
```diff
@@ -550,30 +553,42 @@ fn collate_raw_dylibs(used_libraries: &[NativeLib]) -> Vec<(String, Vec<DllImpor
+        let mut import_table: FxHashMap<Symbol, &DllImport> = FxHashMap::default();
+        for import in imports {
+            if let Some(old_import) = import_table.insert(import.name, &import) {
+                if import != old_import {
```
`DllImport` contains a `Span` now, so you cannot use derived `PartialEq`/`Hash` anymore, spans are almost always going to be different.
`import_table` should probably keep only `DllCallingConvention`s instead of full `DllImport`s.

ricobbe replied:
Ok, it looks like I'm sort of stuck here.  As far as I can determine, something in the query system requires `NativeLib` to implement `HashStable`.  (I'm sort of guessing here because the error happens somewhere in a big stack of macro invocations, and I'm having trouble figuring out where in this stack this requirement is imposed.)  Therefore, adding `Span` information to `NativeLib` (even indirectly, as I have) doesn't seem to be an option.

The `Span` is only necessary so I could give some sort of source location information in the error message that we signal if we encounter multiple declarations of the same external function with inconsistent calling conventions.  Do we want to settle for issuing a diagnostic without location info?  (I'm not wild about this choice.)  Alternatively, is there a better place in the compiler to detect this case and issue the error?  I would expect that a big difficulty with doing this before we've created the `NativeLib` structs is that we don't have easy access to information about all of the external declarations in one place.

> `import_table` should probably keep only `DllCallingConvention`s instead of full `DllImport`s.

I think this gets into a question about how much future proofing we should be doing here.  The RFC does call for eventual support of linking by ordinal instead of by name, and `DllImport` contains the `ordinal` field to support this in the future.  At the moment, nobody ever reads that field.  Going forward, I have no idea if/when I'm going to get around to implementing that specific bit of functionality; our primary concern is to make interoperability with the Windows API easier, and support for ordinals isn't very important for that.  I'm starting to wonder if I should just drop the field (but otherwise leave `DllImport` intact) rather than including only half-hearted support for ordinals.

petrochenkov replied:
>something in the query system requires `NativeLib` to implement `HashStable`

Derived `Hash` implementation (including `Span`) works ok for the purposes of query/incremental systems, it'll just add some dependency on spans, so it can be kept.
The important thing is to not use for logic like duplicate detection here.

>I think this gets into a question about how much future proofing we should be doing here. The RFC does call for eventual support of linking by ordinal instead of by name

Then `FxHashMap<Symbol, DllCallingConvention>` will turn into `FxHashMap<Symbol, (DllCallingConvention, Option<u16>)>`.

petrochenkov commented on compiler/rustc_codegen_ssa/src/back/link.rs:
```diff
@@ -550,30 +553,42 @@ fn collate_raw_dylibs(used_libraries: &[NativeLib]) -> Vec<(String, Vec<DllImpor
+                    sess.span_fatal(
+                        import.span,
+                        &format!(
+                            "multiple inconsistent definitions of external function `{}` from library `{}`",
```
It's better to explicitly say that the functions have different ABIs, otherwise it's not clear how exactly they are inconsistent.

petrochenkov commented:
@ricobbe 
Could you squash commits after addressing the remaining comments?

ricobbe commented:
> Could you squash commits after addressing the remaining comments?

Will do, sure.  For my own understanding, what is preferred practice on this?  Squash at the end before completing the PR, or squash as I go?

petrochenkov commented:
@ricobbe 
>For my own understanding, what is preferred practice on this? Squash at the end before completing the PR, or squash as I go?

Both ways work for me.
The important part is to have a reasonable git history immediately before merging.

ricobbe commented:
@rustbot label -S-waiting-on-author +S-waiting-on-review

petrochenkov commented:
Thanks!
@bors r+

bors commented:
:pushpin: Commit 9c9694f7c748c1a41a3ef9ec524e6d54c667bc56 has been approved by `petrochenkov`

<!-- @bors r=petrochenkov 9c9694f7c748c1a41a3ef9ec524e6d54c667bc56 -->
<!-- homu: {"type":"Approved","sha":"9c9694f7c748c1a41a3ef9ec524e6d54c667bc56","approver":"petrochenkov"} -->

bors commented:
:hourglass: Testing commit 9c9694f7c748c1a41a3ef9ec524e6d54c667bc56 with merge 5d062a65e75e83b6a69930bde9c3b0b86107139b...
<!-- homu: {"type":"BuildStarted","head_sha":"9c9694f7c748c1a41a3ef9ec524e6d54c667bc56","merge_sha":"5d062a65e75e83b6a69930bde9c3b0b86107139b"} -->

rust-log-analyzer commented:
The job **`i686-msvc-2`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/runs/3028760450) [(plain)](https://github.com/rust-lang-ci/rust/commit/5d062a65e75e83b6a69930bde9c3b0b86107139b/checks/3028760450/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
test [ui] ui\wrong-hashset-issue-42918.rs ... ok

failures:

---- [ui] ui\rfc-2627-raw-dylib\multiple-definitions.rs stdout ----
$DIR\multiple-definitions.rs
$DIR\multiple-definitions.rs

8    = note: see issue #58713 <https://github.com/rust-lang/rust/issues/58713> for more information
9 
9 
10 error: multiple definitions of external function `f` from library `foo.dll` have different calling conventions
-   --> $DIR/multiple-definitions.rs:14:9
12    |
12    |
- LL |         fn f(x: i32);
-    |         ^^^^^^^^^^^^^
+ LL |     fn f(x: i32);
15 
16 error: aborting due to previous error; 1 warning emitted
17 



The actual stderr differed from the expected stderr.
Actual stderr saved to D:\a\rust\rust\build\i686-pc-windows-msvc\test\ui\rfc-2627-raw-dylib\multiple-definitions\multiple-definitions.stderr
To update references, rerun the tests and pass the `--bless` flag
To only update this specific test, also pass `--test-args rfc-2627-raw-dylib\multiple-definitions.rs`
error: 1 errors occurred comparing output.
status: exit code: 1
status: exit code: 1
command: PATH="D:\a\rust\rust\build\i686-pc-windows-msvc\stage2\bin;C:\Program Files (x86)\Windows Kits\10\bin\x64;C:\Program Files (x86)\Windows Kits\10\bin\10.0.19041.0\x64;C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\VC\Tools\MSVC\14.29.30037\bin\HostX64\x64;C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\VC\Tools\MSVC\14.29.30037\bin\HostX64\x86;D:\a\rust\rust\build\i686-pc-windows-msvc\stage0-bootstrap-tools\i686-pc-windows-msvc\release\deps;D:\a\rust\rust\build\i686-pc-windows-msvc\stage0\bin;D:\a\rust\rust\ninja;D:\a\rust\rust\msys2\mingw32\bin;C:\hostedtoolcache\windows\Python\3.9.5\x64\Scripts;C:\hostedtoolcache\windows\Python\3.9.5\x64;C:\msys64\usr\bin;D:\a\rust\rust\sccache;C:\Users\runneradmin\.dotnet\tools;C:\Program Files\MongoDB\Server\4.4\bin;C:\aliyun-cli;C:\vcpkg;C:\cf-cli;C:\Program Files (x86)\NSIS;C:\tools\zstd;C:\Program Files\Mercurial;C:\hostedtoolcache\windows\stack\2.7.1\x64;C:\tools\ghc-9.0.1\bin;C:\Program Files\dotnet;C:\mysql-5.7.21-winx64\bin;C:\Program Files\R\R-4.1.0\bin\x64;C:\SeleniumWebDrivers\GeckoDriver;C:\Program Files (x86)\sbt\bin;C:\Rust\.cargo\bin;C:\Program Files (x86)\GitHub CLI;C:\Program Files\Git\bin;C:\Program Files (x86)\pipx_bin;C:\hostedtoolcache\windows\go\1.15.13\x64\bin;C:\hostedtoolcache\windows\Python\3.7.9\x64\Scripts;C:\hostedtoolcache\windows\Python\3.7.9\x64;C:\hostedtoolcache\windows\Ruby\2.5.9\x64\bin;C:\hostedtoolcache\windows\Java_Adopt_jdk\8.0.292-10\x64\bin;C:\npm\prefix;C:\Program Files (x86)\Microsoft SDKs\Azure\CLI2\wbin;C:\ProgramData\kind;C:\Windows\system32;C:\Windows;C:\Windows\System32\Wbem;C:\Windows\System32\WindowsPowerShell\v1.0;C:\Windows\System32\OpenSSH;C:\ProgramData\Chocolatey\bin;C:\Program Files\Microsoft\Web Platform Installer;C:\Program Files\Docker;C:\Program Files\PowerShell\7;C:\Program Files\dotnet;C:\Program Files\Microsoft SQL Server\130\Tools\Binn;C:\Program Files\Microsoft SQL Server\Client SDK\ODBC\170\Tools\Binn;C:\Program Files (x86)\Windows Kits\10\Windows Performance Toolkit;C:\Program Files (x86)\Microsoft SQL Server\110\DTS\Binn;C:\Program Files (x86)\Microsoft SQL Server\120\DTS\Binn;C:\Program Files (x86)\Microsoft SQL Server\130\DTS\Binn;C:\Program Files (x86)\Microsoft SQL Server\140\DTS\Binn;C:\Program Files (x86)\Microsoft SQL Server\150\DTS\Binn;C:\Program Files\nodejs;C:\ProgramData\chocolatey\lib\pulumi\tools\Pulumi\bin;C:\ProgramData\chocolatey\lib\maven\apache-maven-3.8.1\bin;C:\Program Files\Microsoft Service Fabric\bin\Fabric\Fabric.Code;C:\Program Files\Microsoft SDKs\Service Fabric\Tools\ServiceFabricLocalClusterManager;C:\Program Files\OpenSSL\bin;C:\Strawberry\c\bin;C:\Strawberry\perl\site\bin;C:\Strawberry\perl\bin;C:\Program Files\Git\cmd;C:\Program Files\Git\mingw64\bin;C:\Program Files\Git\usr\bin;C:\tools\php;C:\Program Files (x86)\sbt\bin;C:\Program Files\TortoiseSVN\bin;C:\SeleniumWebDrivers\ChromeDriver;C:\SeleniumWebDrivers\EdgeDriver;C:\Program Files\CMake\bin;C:\Program Files\Amazon\AWSCLIV2;C:\Program Files\Amazon\SessionManagerPlugin\bin;C:\Program Files\Amazon\AWSSAMCLI\bin;C:\Program Files (x86)\Google\Cloud SDK\google-cloud-sdk\bin;C:\Program Files (x86)\Microsoft BizTalk Server;C:\Users\runneradmin\AppData\Local\Microsoft\WindowsApps" "D:\\a\\rust\\rust\\build\\i686-pc-windows-msvc\\stage2\\bin\\rustc.exe" "D:\\a\\rust\\rust\\src/test\\ui\\rfc-2627-raw-dylib\\multiple-definitions.rs" "-Zthreads=1" "--target=i686-pc-windows-msvc" "--error-format" "json" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zemit-future-incompat-report" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "D:\\a\\rust\\rust\\build\\i686-pc-windows-msvc\\test\\ui\\rfc-2627-raw-dylib\\multiple-definitions" "-A" "unused" "-Crpath" "-O" "-Cdebuginfo=0" "-Lnative=D:\\a\\rust\\rust\\build\\i686-pc-windows-msvc\\native\\rust-test-helpers" "--crate-type" "lib" "--emit" "link" "-L" "D:\\a\\rust\\rust\\build\\i686-pc-windows-msvc\\test\\ui\\rfc-2627-raw-dylib\\multiple-definitions\\auxiliary"
------------------------------------------

------------------------------------------
stderr:
stderr:
------------------------------------------
warning: the feature `raw_dylib` is incomplete and may not be safe to use and/or cause compiler crashes
  --> D:\a\rust\rust\src/test\ui\rfc-2627-raw-dylib\multiple-definitions.rs:4:12
   |
LL | #![feature(raw_dylib)]
   |
   = note: `#[warn(incomplete_features)]` on by default
   = note: see issue #58713 <https://github.com/rust-lang/rust/issues/58713> for more information


error: multiple definitions of external function `f` from library `foo.dll` have different calling conventions
  --> D:\a\rust\rust\src/test\ui\rfc-2627-raw-dylib\multiple-definitions.rs:8:5
   |
LL |     fn f(x: i32);

error: aborting due to previous error; 1 warning emitted


---
    [ui] ui\rfc-2627-raw-dylib\multiple-definitions.rs

test result: FAILED. 11888 passed; 1 failed; 183 ignored; 0 measured; 0 filtered out; finished in 354.57s

Some tests failed in compiletest suite=ui mode=ui host=i686-pc-windows-msvc target=i686-pc-windows-msvc


command did not execute successfully: "D:\\a\\rust\\rust\\build\\i686-pc-windows-msvc\\stage0-tools-bin\\compiletest.exe" "--compile-lib-path" "D:\\a\\rust\\rust\\build\\i686-pc-windows-msvc\\stage2\\bin" "--run-lib-path" "D:\\a\\rust\\rust\\build\\i686-pc-windows-msvc\\stage2\\lib\\rustlib\\i686-pc-windows-msvc\\lib" "--rustc-path" "D:\\a\\rust\\rust\\build\\i686-pc-windows-msvc\\stage2\\bin\\rustc.exe" "--src-base" "D:\\a\\rust\\rust\\src/test\\ui" "--build-base" "D:\\a\\rust\\rust\\build\\i686-pc-windows-msvc\\test\\ui" "--stage-id" "stage2-i686-pc-windows-msvc" "--suite" "ui" "--mode" "ui" "--target" "i686-pc-windows-msvc" "--host" "i686-pc-windows-msvc" "--llvm-filecheck" "D:\\a\\rust\\rust\\build\\i686-pc-windows-msvc\\llvm\\build\\bin\\FileCheck.exe" "--nodejs" "C:\\Program Files\\nodejs\\node" "--npm" "C:\\Program Files\\nodejs\\npm" "--host-rustcflags" "-Crpath -O -Cdebuginfo=0  -Lnative=D:\\a\\rust\\rust\\build\\i686-pc-windows-msvc\\native\\rust-test-helpers" "--target-rustcflags" "-Crpath -O -Cdebuginfo=0  -Lnative=D:\\a\\rust\\rust\\build\\i686-pc-windows-msvc\\native\\rust-test-helpers" "--docck-python" "C:\\hostedtoolcache\\windows\\Python\\3.9.5\\x64\\python3.exe" "--lldb-python" "C:\\hostedtoolcache\\windows\\Python\\3.9.5\\x64\\python3.exe" "--gdb" "C:\\msys64\\usr\\bin\\gdb" "--llvm-version" "12.0.1-rust-1.55.0-nightly" "--llvm-components" "aarch64 aarch64asmparser aarch64codegen aarch64desc aarch64disassembler aarch64info aarch64utils aggressiveinstcombine all all-targets analysis arm armasmparser armcodegen armdesc armdisassembler arminfo armutils asmparser asmprinter avr avrasmparser avrcodegen avrdesc avrdisassembler avrinfo binaryformat bitreader bitstreamreader bitwriter bpf bpfasmparser bpfcodegen bpfdesc bpfdisassembler bpfinfo cfguard codegen core coroutines coverage debuginfocodeview debuginfodwarf debuginfogsym debuginfomsf debuginfopdb demangle dlltooldriver dwarflinker engine executionengine extensions filecheck frontendopenacc frontendopenmp fuzzmutate globalisel hellonew hexagon hexagonasmparser hexagoncodegen hexagondesc hexagondisassembler hexagoninfo instcombine instrumentation interfacestub interpreter ipo irreader jitlink libdriver lineeditor linker lto mc mca mcdisassembler mcjit mcparser mips mipsasmparser mipscodegen mipsdesc mipsdisassembler mipsinfo mirparser msp430 msp430asmparser msp430codegen msp430desc msp430disassembler msp430info native nativecodegen nvptx nvptxcodegen nvptxdesc nvptxinfo objcarcopts object objectyaml option orcjit orcshared orctargetprocess passes powerpc powerpcasmparser powerpccodegen powerpcdesc powerpcdisassembler powerpcinfo profiledata remarks riscv riscvasmparser riscvcodegen riscvdesc riscvdisassembler riscvinfo runtimedyld scalaropts selectiondag sparc sparcasmparser sparccodegen sparcdesc sparcdisassembler sparcinfo support symbolize systemz systemzasmparser systemzcodegen systemzdesc systemzdisassembler systemzinfo tablegen target textapi transformutils vectorize webassembly webassemblyasmparser webassemblycodegen webassemblydesc webassemblydisassembler webassemblyinfo windowsmanifest x86 x86asmparser x86codegen x86desc x86disassembler x86info xray" "--cc" "" "--cxx" "" "--cflags" "" "--adb-path" "adb" "--adb-test-dir" "/data/tmp/work" "--android-cross-path" "" "--channel" "nightly" "--color" "always"


Build completed unsuccessfully in 0:33:33
Build completed unsuccessfully in 0:33:33
make: *** [Makefile:74: ci-subset-2] Error 1
```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/3029044715)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/3029044715","builder_name":"checks-actions"} -->

petrochenkov commented on compiler/rustc_codegen_ssa/src/back/link.rs:
```diff
@@ -464,30 +467,44 @@ fn collate_raw_dylibs(used_libraries: &[NativeLib]) -> Vec<(String, Vec<DllImpor
+    // warning.
+    for (library, imports) in &dylib_table {
+        let mut import_table: FxHashMap<Symbol, DllCallingConvention> = FxHashMap::default();
+        for import in imports {
```
"multiple definitions" are reported by walking an `FxHashSet`, so the walking order is unpredictable and `.stderr` files in the test suite can be different causing the CI failure.

Perhaps this logic needs to be placed after the sorting code below.

ricobbe replied:
Yup, just figured this out.  Working on a fix now; should only take an hour or so, barring interrupts from other stuff.

ricobbe commented:
@rustbot label -S-waiting-on-author +S-waiting-on-review

petrochenkov commented:
@bors r+

bors commented:
:pushpin: Commit a867dd4c7e65beb1453fcff9e252e7ab19f0e1d0 has been approved by `petrochenkov`

<!-- @bors r=petrochenkov a867dd4c7e65beb1453fcff9e252e7ab19f0e1d0 -->
<!-- homu: {"type":"Approved","sha":"a867dd4c7e65beb1453fcff9e252e7ab19f0e1d0","approver":"petrochenkov"} -->

bors commented:
:hourglass: Testing commit a867dd4c7e65beb1453fcff9e252e7ab19f0e1d0 with merge 8d9d4c87d677552ae52e2d58034e4be199b5a6d2...
<!-- homu: {"type":"BuildStarted","head_sha":"a867dd4c7e65beb1453fcff9e252e7ab19f0e1d0","merge_sha":"8d9d4c87d677552ae52e2d58034e4be199b5a6d2"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/3034031883)
Approved by: petrochenkov
Pushing 8d9d4c87d677552ae52e2d58034e4be199b5a6d2 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"petrochenkov","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/3034031883"},"merge_sha":"8d9d4c87d677552ae52e2d58034e4be199b5a6d2"} -->

joshtriplett commented:
This appears to have added a failing test:
```
8    = note: see issue #58713 <https://github.com/rust-lang/rust/issues/58713> for more information
9 
9 
10 error: multiple definitions of external function `f` from library `foo.dll` have different calling conventions
-   --> $DIR/multiple-definitions.rs:8:5
12    |
12    |
- LL |     fn f(x: i32);
-    |     ^^^^^^^^^^^^^
+ LL |         fn f(x: i32);
15 
16 error: aborting due to previous error; 1 warning emitted
17
```

RalfJung commented:
> This appears to have added a failing test:

How is that even possible...?

fee1-dead commented:
This is weird, both #86922 and #86857 encountered this but they seem unrelated.

#87075 *fails* again.

RalfJung commented:
Looks like we have different output on different targets?
But bors wouldn't have landed unless the stderr file is correct for *all* targets...

RalfJung commented:
Is it possible that something depends on hash table ordering here (like the code that determines which of the "multiple definitions" the span is pointing at)?

ricobbe commented:
I'm finding it a bit hard to follow the flurry of communications about this PR that happened over the weekend.  What's the current status of this work--is the multiple-definitions test still failing in automation?

> Is it possible that something depends on hash table ordering here (like the code that determines which of the "multiple definitions" the span is pointing at)?

An earlier version of this PR did indeed have this problem.  I checked in a fix before the PR was merged, but it's possible that there's still some nondeterminism left.
